### PR TITLE
[#337] Allow continuous aliveness checkers

### LINKED ISSUE #59: Battery checking should be continuous, not just at steps
**Describe the bug**
The satellite can die then come back alive over the course of a step without triggering failure.

Screenshot 2023-09-13 at 12 08 22 PM

**Additional context**
Should be fixed by either logging battery and checking at the end of the step, or by adding a continuous terminal event checking battery level. I think the latter is preferable, as it will be less memory intensive (or maybe not, turns out events are time-expensive to check.)

### LINKED ISSUE #337: Aliveness checkers should be easily addable as per-integrator-step checks using the event system
Currently, aliveness checkers only operate at environment steps; some make sense as continuous checks that interrupt the environment. These could be implemented using Basilisk events.

### PR DESCRIPTION
**Pull request by Strtus**
## Description
Closes #337 
Closes #59

Adds a `continuous` option to `@aliveness_checker` so checkers can run as Basilisk events and stop the environment step when they fail. Existing checkers are unchanged.

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### How should this pull request be reviewed?
- [ ] By commit
- [X] All changes at once

## How Has This Been Tested?

Unit tests for the decorator and event setup. Integration test comparing step-level vs continuous interruption.

## Future Work

Built-in checkers (battery, altitude, RW speed, etc.) still only run at environment steps.

## Checklist

- [X] I have performed a self-review of my code
- [X] I have commented my code in hard-to-understand areas
- [X] I have made corresponding changes to the documentation and release notes
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/bsk_rl/sats/satellite.py [hunk 1]
```diff
@@ -258,6 +259,45 @@ def set_fsw(self, fsw_rate: float) -> "fsw.FSWModelABC":
...
+                    checker = getattr(model, name)
+                    check_rate = checker.check_rate
+                    if check_rate is None:
+                        check_rate = self.simulator.sim_rate
+
+                    def condition(sim, checker=checker):
+                        return not checker()
+
+                    def side_effect(sim, name=name):
+                        self.logger.warning(f"failed {name} check")
```
Comment (Yume27) [inline src/bsk_rl/sats/satellite.py:285]: The continuous event path logs manually in side_effect, while step-based aliveness failures log through aliveness_checker(..., log_failure=True). Is that intentional? It may be worth centralizing the failure logging or preserving the log_failure pattern so continuous and non-continuous checker behavior stays consistent.

File: tests/unittest/utils/test_functional.py [hunk 2]
```diff
@@ -125,6 +126,22 @@ def test_dead(self, type):
         assert functional.check_aliveness_checkers(d, log_failure=True) is False
         d.satellite.logger.warning.assert_called_with("failed is_living check")
 
+    def test_continuous_defaults(self):
```
Comment (Yume27) [inline tests/unittest/utils/test_functional.py:129]: Could you add coverage for the empty-parentheses decorator form, @functional.aliveness_checker()? That exercises the new func is None branch with default options.

File: tests/integration/sim/test_int_dynamics.py [hunk 3]
```diff
@@ -240,3 +241,34 @@ class DeputySat(sats.Satellite):
             assert sat2.dynamics.out_of_ranges == []
             assert not terminated["Chief"]
             assert not terminated["Deputy"]
+
+
+class TestContinuousAlivenessChecker:
```
Comment (Yume27) [inline tests/integration/sim/test_int_dynamics.py:246]: The current integration test covers a continuous aliveness checker on the dynamics model. Could we add a similar test for an FSW-defined checker to verify that path also interrupts the environment step as expected?

File: tests/unittest/sats/test_satellite.py [hunk 4]
```diff
@@ -143,6 +146,64 @@ def test_satellite_command(self):
...
+
+        assert sat.simulator.createNewEvent.call_count == 2
+        event_names = [
+            call.args[0] for call in sat.simulator.createNewEvent.call_args_list
+        ]
+        assert any("continuous_valid" in name and "Dyn" in name for name in event_names)
+        assert any("fsw_valid" in name and "Fsw" in name for name in event_names)
+        assert not any("ProxyType" in name for name in event_names)
+
+        dyn_call = sat.simulator.createNewEvent.call_args_list[0]
```
Comment (Yume27) [inline tests/unittest/sats/test_satellite.py:194]: Could you add FSW aliveness_checker tests like dyn_call and also assert the FSW event rate when check_rate=None? It should default to the simulator rate.

File: tests/unittest/sats/test_satellite.py [hunk 5]
```diff
@@ -143,6 +146,64 @@ def test_satellite_command(self):
...
+        assert dyn_call.kwargs["terminal"] is True
+        assert dyn_call.kwargs["exactRateMatch"] is False
+
+        assert dyn_call.kwargs["conditionFunction"](sat.simulator) is False
+        sat.dynamics.fail_continuous = True
+        assert dyn_call.kwargs["conditionFunction"](sat.simulator) is True
+        dyn_call.kwargs["actionFunction"](sat.simulator)
+        assert sat._is_alive is False
+        assert sat.time_of_death == 12.0
+
```
Comment (Yume27) [inline tests/unittest/sats/test_satellite.py:206]: Could you add an assertion for the warning emitted by the continuous event action?

File: tests/integration/sim/test_int_dynamics.py [hunk 6]
```diff
@@ -240,3 +241,34 @@ class DeputySat(sats.Satellite):
...
+    @pytest.mark.parametrize("continuous,expected_time", [(False, 100.0), (True, 5.0)])
+    def test_step_vs_continuous(self, continuous, expected_time):
+        class BoomDyn(dyn.DynamicsModel):
+            @aliveness_checker(continuous=continuous)
```
Comment (Yume27) [inline tests/integration/sim/test_int_dynamics.py:250]: Could you test a few explicit check_rate values in addition to None? That would verify the registered event cadence is wired through correctly.

**Review by Yume27 — CHANGES_REQUESTED** [review]
Thank you for adding this functionality. This helps make the simulation behavior more realistic for operations where failures can occur during a long environment step. That said, I think some existing aliveness checkers would benefit from using this behavior by default. In particular, `battery_valid` and `rw_speeds_valid` seem like strong candidates for `continuous=True`, since transient battery depletion or reaction wheel overspeed should not be missed if they occur mid-step. Could we update those checkers to use continuous aliveness checking by default?

**Comment by Strtus** [comment]
Thanks for the review @Yume27  I’ve been busy these past few days and haven’t had time to work through the comments yet. I’ll get back to this and update the PR in the next couple of days.
